Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -0,0 +1,215 @@
+                try writer.truncate(atOffset: writer.offsetInFile)
+
+                // Set the offset back to the beginning of the file.
+                try writer.seek(toOffset: 0)
+
+                // We know the oldest message is at the beginning of the file, since we just tossed out the rest of the file.
+                try reader.seek(toOffset: 0)
+
+                // We know we're about to overwrite the oldest message, so advance the reader to the second oldest message.
+                try reader.seekToNextMessage(shouldSeekToOldestMessageIfFound: false, cacheOverwritesOldMessages: true)
+            }
+
+            // Prepare the reader before writing the message.
+            try prepareReaderForWriting(dataOfLength: bytesNeededToStoreMessage)
+
+            // Create the marker for the offset representing the beggining of the message that will be the oldest once our write is done.
+            let offsetInFileOfOldestMessage = Data(Bytes(reader.offsetInFile))
+
+            // Write the message and oldest message offset.
+            try write(messageData: messageData, offsetInFileOfOldestMessage: offsetInFileOfOldestMessage)
+
+        } else if cacheHasSpaceForNewMessageBeforeEndOfFile {
+            // Write the message.
+            try write(messageData: messageData, offsetInFileOfOldestMessage: nil)
+
+        } else {
+            // We're out of room.
+            throw CacheAdvanceWriteError.messageDataTooLarge
+        }
+    }
+
+    /// Fetches all messages from the cache.
+    public func messages() throws -> [T] {
+        try setUpFileHandlesIfNecessary()
+
+        var messages = [T]()
+        while let encodedMessage = try reader.nextEncodedMessage(cacheOverwritesOldMessages: shouldOverwriteOldMessages) {
+            messages.append(try decoder.decode(T.self, from: encodedMessage))
+        }
+
+        return messages
+    }
+
+    // MARK: Private
+
+    /// Seeks the reader and writer file handles to the correct place.
+    /// Only necessary the first time this method is called.
+    private func setUpFileHandlesIfNecessary() throws {
+        guard !hasSetUpFileHandles else {
+            return
+        }
+
+        // This is our first cache action.
+        // We need to find out where we should write our next message.
+        try reader.seekToEndOfNewestMessage(cacheOverwritesOldMessages: shouldOverwriteOldMessages)
+        try writer.seek(toOffset: reader.offsetInFile)
+
+        // Now that we know where to write, we need to figure out where the oldest message is.
+        try reader.seekToBeginningOfOldestMessage(cacheOverwritesOldMessages: shouldOverwriteOldMessages)
+
+        hasSetUpFileHandles = true
+    }
+
+    /// Advances the reader until there is room to store a new message without writing past the reader.
+    /// - Parameter messageLength: the length of the next message that will be written.
+    private func prepareReaderForWriting(dataOfLength messageLength: Bytes) throws {
```

Comment:
I'll add documentation. The reader is basically useless (outside of reading all messages) in a cache that doesn't overwrite old messages.

I'm musing here, but I'm curious how much benefit we're getting from having a single cache type that can either overwrite old messages and one that throws when it is full in a single class.

Maybe it's worth splitting out the functionality into a `CacheAdvance` and an `OverwritingCacheAdvance`? (This would be a separate PR)